pkg/agent/cniserver/interface_configuration_linux.go

```diff
@@ -91,49 +92,154 @@ func (ic *ifConfigurator) moveIfToNetns(ifname string, netns ns.NetNS) error {
+	tempNS, err := tempNetNS()
+	if err != nil {
+		return fmt.Errorf("failed to create tempNS: %w", err)
+	}
+	defer tempNS.Close()
+
+	// Move the host VF device into tempNS
+	if err = ic.netlink.LinkSetNsFd(link, int(tempNS.Fd())); err != nil {
+		return fmt.Errorf("failed to move %s to tempNS: %w", vfNetDevice, err)
+	}
+
+	netns, err := nsGetNS(containerNetNS)
+	if err != nil {
+		return fmt.Errorf("failed to open container netns %s: %w", containerNetNS, err)
+	}
+	defer netns.Close()
+
+	if err = tempNS.Do(func(hostNS ns.NetNS) error {
+		// Look up the device in tempNS, as the index might have changed
+		tempNSDev, err := ic.netlink.LinkByName(vfNetDevice)
 		if err != nil {
-			return fmt.Errorf("failed to set MTU for VF netdevice %s: %v", containerIfaceName, err)
+			return fmt.Errorf("failed to find %s in tempNS: %w", vfNetDevice, err)
 		}
-		err = ic.netlink.LinkSetUp(link)
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				_ = ic.netlink.LinkSetNsFd(tempNSDev, int(hostNS.Fd()))
+			}
+		}()
+
+		// Rename the device to the wanted name
+		if err = ic.netlink.LinkSetName(tempNSDev, containerIfaceName); err != nil {
+			return fmt.Errorf("failed to rename VF device %s to %s: %w", vfNetDevice, containerIfaceName, err)
+		}
+
+		// Restore the original device name in case of error
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				_ = ic.netlink.LinkSetName(tempNSDev, vfNetDevice)
+			}
+		}()
+
+		err = ic.netlink.LinkSetAlias(tempNSDev, vfNetDevice)
 		if err != nil {
-			return fmt.Errorf("failed to set link up to VF netdevice %s: %v", containerIfaceName, err)
+			return fmt.Errorf("failed to set alias as %s for VF netdevice %s: %w", vfNetDevice, vfNetDevice, err)
 		}
-		containerIface.Name = containerIfaceName
-		containerIface.Mac = link.Attrs().HardwareAddr.String()
-		containerIface.Sandbox = netns.Path()
-		klog.V(2).Infof("hostIface: %+v, containerIface: %+v", hostIface, containerIface)
-		klog.V(2).Infof("Configuring IP address for container %s", containerID)
-		// result.Interfaces must be set before this.
-		if err := ipamConfigureIface(containerIface.Name, result); err != nil {
-			return fmt.Errorf("failed to configure IP address for container %s: %v", containerID, err)
+		klog.V(2).InfoS("Link's alias has been updated", "alias", vfNetDevice)
+
+		// Remove the alias on error
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				_ = ic.netlink.LinkSetAlias(tempNSDev, "")
+			}
+		}()
+
+		// Move VF to Container namespace
+		err = ic.moveIFToNetNS(containerIfaceName, netns)
+		if err != nil {
+			return fmt.Errorf("failed to move VF %s to container netns %s: %w", vfNetDevice, containerNetNS, err)
 		}
-		klog.V(2).Infof("ipam.ConfigureIface result: %+v, err: %v", result, err)
-		return nil
+
+		// Look up the device again on error, as the index might have changed
+		defer func() {
+			if err != nil {
+				tempNSDev, _ = ic.netlink.LinkByName(containerIfaceName)
```

Comment:
The interface can be moved to some other NS already?

Reply:
Not sure I got the question, we moved the interface to the expected container NS before this defer function. But this step is after the netns is moved to tempNS in the defer function chains. This step is trying to get the right link and continue the undo actions like reset VF interface name and empty the alias etc.

Reply:
I meant when the interface is not in the root NS, could ic.netlink.LinkByName(containerIfaceName) still find the interface by name?

And should we check if ic.netlink.LinkByName(containerIfaceName) returns error or not here?

Reply:
This step is actually still inside of 'tempNS.Do(func(hostNS ns.NetNS)',  not in root NS. 
Regarding the returned error, it's also discarded in host-device, we may record the error as a reference, but it's probably hard to do anything unless we want to introduce retry mechanism in every defer step.

Reply:
Updated with a few logs to print errors in defer steps.